File: data_breach_analysis.go
package main

import (
	"fmt"
	"strings"
)

// Function to analyze logs for breach events
func analyzeLogs(logs []string) (int, string) {
	var breachCount int
	var report strings.Builder

	report.WriteString("Data Breach Analysis Report for Singapore Server\n")
	report.WriteString("------------------------------------------------\n")
	report.WriteString("Summary: This report analyzes server logs for potential data breaches.\n\n")

	for i, logEntry := range logs {
		if strings.Contains(strings.ToLower(logEntry), "breach") || strings.Contains(strings.ToLower(logEntry), "unauthorized access") {
			breachCount++
			report.WriteString(fmt.Sprintf("Event %d: Potential breach detected - %s\n", i+1, logEntry))
		}
	}

	report.WriteString(fmt.Sprintf("\nTotal potential breach events: %d\n", breachCount))
	if breachCount > 0 {
		report.WriteString("Recommendation: Investigate immediately and notify authorities.")
	} else {
		report.WriteString("No breach events detected. System appears secure.")
	}

	return breachCount, report.String()
}

func main() {
	// Sample logs for Singapore server (in a real scenario, read from a file)
	sampleLogs := []string{
		"2023-10-01 12:00:00 - Normal login from user123",
		"2023-10-01 12:05:00 - Unauthorized access attempt on Singapore server",
		"2023-10-01 12:10:00 - Breach detected in database",
		"2023-10-01 12:15:00 - System restored",
	}

	breachCount, report := analyzeLogs(sampleLogs)
	fmt.Println(report)
	fmt.Printf("Analysis complete. Total breaches: %d\n", breachCount)
}

File: data_breach_analysis.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>  // For std::transform

// Function to analyze logs for breach events
int analyzeLogs(const std::vector<std::string>& logs, std::string& report) {
    int breachCount = 0;
    report = "Data Breach Analysis Report for Singapore Server\n";
    report += "------------------------------------------------\n";
    report += "Summary: This report analyzes server logs for potential data breaches.\n\n";

    for (size_t i = 0; i < logs.size(); ++i) {
        std::string lowerEntry = logs[i];  // Convert to lowercase for case-insensitive search
        std::transform(lowerEntry.begin(), lowerEntry.end(), lowerEntry.begin(), ::tolower);
        
        if (lowerEntry.find("breach") != std::string::npos || lowerEntry.find("unauthorized access") != std::string::npos) {
            breachCount++;
            report += "Event " + std::to_string(i + 1) + ": Potential breach detected - " + logs[i] + "\n";
        }
    }

    report += "\nTotal potential breach events: " + std::to_string(breachCount) + "\n";
    if (breachCount > 0) {
        report += "Recommendation: Investigate immediately and notify authorities.";
    } else {
        report += "No breach events detected. System appears secure.";
    }

    return breachCount;
}

int main() {
    // Sample logs for Singapore server (in a real scenario, read from a file)
    std::vector<std::string> sampleLogs = {
        "2023-10-01 12:00:00 - Normal login from user123",
        "2023-10-01 12:05:00 - Unauthorized access attempt on Singapore server",
        "2023-10-01 12:10:00 - Breach detected in database",
        "2023-10-01 12:15:00 - System restored"
    };

    std::string report;
    int breachCount = analyzeLogs(sampleLogs, report);
    
    std::cout << report << std::endl;
    std::cout << "Analysis complete. Total breaches: " << breachCount << std::endl;

    return 0;
}

// To handle file I/O in a real scenario:
// #include <fstream>
// std::vector<std::string> logs;
// std::ifstream file("server.log");
// std::string line;
// while (std::getline(file, line)) {
//     logs.push_back(line);
// }
// Then call analyzeLogs(logs, report);
